### PR TITLE
Refactor AST (Big one) for convenience

### PR DESCRIPTION
**Pull request by TilakMaddy**
**Refactors**

* Separate the AST methods into different sub folders - 
1. `own`  -  Contains the direct implementations + implementations of std library methods like (partialEq, hash) on the node
2. `node` - Contains the `impl Node for ...` methods (`Node` is our custom defined trait)
3. `disp` - Contains the `impl Display for ...` methods  (`Display` is not a custom defined trait, but it's an exception)

* All the AST Node structs are defined  in the same file `ast_nodes.rs`
(Idea borrowed from foundry compilers)

* Convert Node Structs to macro representation 

* Capturable methods generated by macros

* **🧽 Cleaned up workspace context**

* Added IDs and src for TypeName node variants, Emit Statements etc 

* Fix ForLoop initialization variants

* Macros for generating extraction library 

These macros will help us easily manage code a lot easier. Macros are responsible for implementing the methods, so no human error and ensures consistent logic.  Another advantage is we can update logic in one place to update everything

Plus, navigate and think about building better abstractions on ASTs without looking at many different files to get an idea of something (for example: what FunctionCall looks like?)

UPDATE:

Thanks to doing this, I found some issues in the AST now, that I couldn't before. So it will require more work but that is outside the scope of refactoring ...So it will be done in a different PR

UPDATE AGAIN:

Making some fixes in this PR only was inevitable. For example, now all AST nodes (other than yul) have ID accessible and put into the context

### CODE REVIEW

**Comment by TilakMaddy** [comment]
Follow Up PR: propagate the `accept()` calls from the newly supported nodes that have IDs

**Comment by TilakMaddy** [comment]
Removed a total of ~2000 lines of code with macros : )

**Comment by TilakMaddy** [comment]
@alexroan Can you take a look at this before working on/reviewing other PRs ?

**Comment by alexroan** [comment]
> @alexroan Can you take a look at this before working on/reviewing other PRs ?

Damn, sorry @TilakMaddy I've just seen this after merging a previous PR.

I will review this regardless, but I think you'll have to update this branch from dev and fix any fails.

File: aderyn_core/src/context/meta_workspace.rs [hunk 1]
```diff
@@ -0,0 +1,239 @@
+use crate::ast::*;
```
Comment (alexroan) [inline aderyn_core/src/context/meta_workspace.rs:1]: What's the difference between workspace and meta_workspace? Why the separation?

**Review by alexroan — COMMENTED** [review]
Wow, @TilakMaddy  this is a big one, but it looks like some epic use of macros (which scares me a little).

I'm not fully confident with macros and rules, so I've added loads of questions here to try to understand more.

Thanks dude, this looks epic.

**Comment by alexroan** [comment]
Hey @TilakMaddy , I'm going to mark this as going into 0.1.6.

It's a big enough structural change to deserve its own release version.

File: aderyn_core/src/ast/impls/own/contracts.rs [hunk 2]
```diff
@@ -1,75 +1,4 @@
-use super::*;
```
[the remark this answers is not in the document]
Comment (TilakMaddy) [reply aderyn_core/src/ast/impls/own/contracts.rs:1]: Could you suggest please? I also didn't like `own` very much 😅 I thought of `self` first but then it's a keyword

File: aderyn_core/src/context/meta_workspace.rs (hunk 1 above)
Comment (TilakMaddy) [reply aderyn_core/src/context/meta_workspace.rs:1]: There is No difference. The file was getting too big so I thought I'll put it in a `meta` file :P

File: aderyn_core/src/context/browser/macros.rs [hunk 3]
```diff
@@ -0,0 +1,27 @@
+macro_rules! generate_extraction_library {
```
[the remark this answers is not in the document]
Comment (TilakMaddy) [reply aderyn_core/src/context/browser/macros.rs:1]: 💪

File: aderyn_core/src/ast/macros.rs [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+                ),*
+            }
+        );
+    }
+}
+
+/// A macro that expands to an enum where each variant also contains a struct of the same name.
+///
+/// The inner value of each variant is boxed since AST types are inherently recursive.
+macro_rules! node_group {
```
[the remark this answers is not in the document]
Comment (TilakMaddy) [reply aderyn_core/src/ast/macros.rs:88]: Thanks for the catch. It's not used. I'll remove it

File: aderyn_core/src/ast/macros.rs [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+                $(#[$field_meta:meta])*
+                $field:ident: $ty:ty
+            ),* $(,)?
+        }
+    ) => {
+        $(#[$struct_meta])*
+        #[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+        #[serde(rename_all = "camelCase")]
+        pub struct $name {
+            pub id: $crate::ast::NodeID,
```
[the remark this answers is not in the document]
Comment (TilakMaddy) [reply aderyn_core/src/ast/macros.rs:18]: Yes ser !

File: aderyn_core/src/ast/macros.rs [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+            pub src: String,
+            $(
+                $(#[$field_meta])*
+                pub $field: $ty
+            ),*
+        }
+    };
+}
+
+macro_rules! ast_node_no_partial_eq {
```
[the remark this answers is not in the document]
Comment (TilakMaddy) [reply aderyn_core/src/ast/macros.rs:28]: I'll think about it and let you... For now, I have only "translated" the existing code to macros. No new logic from my side. So the ones which `PartialEq` already carried on here, and the ones that didn't have been dropped.

File: aderyn_core/src/ast/macros.rs [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+            pub src: String,
+            $(
+                $(#[$field_meta])*
+                pub $field: $ty
+            ),*
+        }
+    };
+}
+
+/// A macro that expands to a struct with common expression node fields.
```
[the remark this answers is not in the document]
Comment (TilakMaddy) [reply aderyn_core/src/ast/macros.rs:52]: Every expression node *is* as ast node, but with extra following parameters
```
                argument_types: Option<Vec<TypeDescriptions>>,
                is_constant: bool,
                is_l_value: bool,
                is_pure: bool,
                l_value_requested: bool,
                type_descriptions: TypeDescriptions,
```

[the remark this answers is not in the document]
Comment (TilakMaddy) [reply aderyn_core/src/ast/macros.rs:52]: Now, later I'm also going to factor out statement nodes from ast_nodes, then based on the ones that remain, I can tell you what an ast_node really is. (For now, think of it like every non expression that is also not a statement, ... (something that doesn't make sense by itself) is an ast_node)

[the remark this answers is not in the document]
Comment (TilakMaddy) [reply aderyn_core/src/ast/macros.rs:52]: Right now, all these variants are `expr_node`

pub enum Expression {
    Literal(Literal),
    Identifier(Identifier),
    UnaryOperation(UnaryOperation),
    BinaryOperation(BinaryOperation),
    Conditional(Conditional),
    Assignment(Assignment),
    FunctionCall(FunctionCall),
    FunctionCallOptions(FunctionCallOptions),
    IndexAccess(IndexAccess),
    IndexRangeAccess(IndexRangeAccess),
    MemberAccess(MemberAccess),
    ElementaryTypeNameExpression(ElementaryTypeNameExpression),
    TupleExpression(TupleExpression),
    NewExpression(NewExpression),
}

**Pull request closed without merging**